src/cr/cube/crunch_cube.py

```diff
@@ -200,82 +200,189 @@ def _as_array(self, include_missing=False, get_non_selected=False,
-            elif len(self.dimensions) == 3:
-                return 2
-            return 1
-
-        inserted_inds = self.inserted_hs_indices() if transforms else []
-        inserted_rows = np.array(inserted_inds[0] if len(inserted_inds) else [])
-        axis = _get_axis_for_margin_calculations()
-        row_margin = self.margin(include_transforms_for_dims=transforms, axis=axis)
-        ind_inserted = np.zeros(row_margin.shape, dtype=bool)
-        if any(inserted_rows):
-            ind_inserted[inserted_rows] = True
-        row_prune = np.logical_or(row_margin == 0, np.isnan(row_margin))
-        row_prune = np.logical_and(row_prune, ~ind_inserted)
-
-        return row_prune
-
-    def _col_prune_indices(self, transforms=None):
-        inserted_inds = self.inserted_hs_indices() if transforms else []
-        inserted_cols = np.array(inserted_inds[1] if len(inserted_inds) > 1 else [])
-        axis = 0 if len(self.dimensions) < 3 else 1
-        col_margin = self.margin(include_transforms_for_dims=transforms, axis=axis)
-
-        # don't consider insertions outside the scope of the columns
-        inserted_cols = inserted_cols[inserted_cols < len(col_margin)]
-
-        ind_inserted = np.zeros(col_margin.shape, dtype=bool)
-        if any(inserted_cols):
-            ind_inserted[inserted_cols] = True
-        col_prune = np.logical_or(col_margin == 0, np.isnan(col_margin))
-        col_prune = np.logical_and(col_prune, ~ind_inserted)
-
-        return col_prune
+    def _prune_3d_indices(self, transforms=None):
+        row_margin = self.margin(
+            include_transforms_for_dims=transforms,
+            axis=self.row_direction_axis
+        )
+        col_margin = self.margin(
+            include_transforms_for_dims=transforms,
+            axis=self.col_direction_axis,
+        )
+        row_inserted_indices = (
+            self.inserted_rows_inds if transforms else []
+        )
+        col_inserted_indices = (
+            self.inserted_col_inds if transforms else []
+        )
+        prune_indices = []
+
+        # It's necessary to go with a for loop, since we don't know if it's
+        # going to be row or a colunm (and across which axis we might
+        # need to sum, in case of 2D margins).
+        for rm, cm in zip(row_margin, col_margin):
+            if rm.ndim > 1:
+                rm = np.sum(rm, axis=1)
+            if cm.ndim > 1:
+                cm = np.sum(cm, axis=0)
+
+            prune_indices.append((
+                self._margin_pruned_indices(rm, row_inserted_indices),
+                self._margin_pruned_indices(cm, col_inserted_indices),
+            ))
+        return prune_indices
+
+    @property
+    def row_direction_axis(self):
+        # when dealing with 1D MR cubes, axis for margin should be 0
+        if len(self.dimensions) == 1 and self.has_mr:
+            return 0
+        elif len(self.dimensions) == 3 and not self.is_double_mr:
+            return 2
```

Comment:
What happens if this is a `double_mr` and also 3d (eg cat x mr x mr)?

Reply:
This will probably _not_ work in that case, but in lack of a decent test case, I left it as is. Maybe we can ticket it (as I'd really like to turn this in). Or if it's an easy case to create, let's go with that. Would you be able to help with that @jonkeane ?

Reply:
I've got an example cube response with this shape, but agree that that alone shouldn't hold this PR up.

Reply:
There's also this that already exists in the repo as well: https://github.com/Crunch-io/crunch-cube/blob/master/tests/integration/fixtures/cubes/cat-x-mr-x-mr.json